nomos-services/chain-service/src/processor.rs

```diff
@@ -0,0 +1,267 @@
+    Cryptarchia, Error, SamplingRelay,
+};
+
+/// Defines how to validate and apply blocks to the [`Cryptarchia`].
+#[async_trait::async_trait]
+pub trait BlockProcessor<RuntimeServiceId> {
+    type Block;
+
+    async fn process_block<CryptarchiaState, Storage>(
+        &mut self,
+        cryptarchia: Cryptarchia<CryptarchiaState>,
+        block: Self::Block,
+        storage: &mut Storage,
+    ) -> Result<Cryptarchia<CryptarchiaState>, (Error, Cryptarchia<CryptarchiaState>)>
+    where
+        CryptarchiaState: cryptarchia_engine::CryptarchiaState + Send,
+        Storage: StorageAdapter<RuntimeServiceId, Block = Self::Block>
+            + StorageAdapterExt<RuntimeServiceId>
+            + Send
+            + Sync
+            + 'static;
+}
+
+/// Implements [`BlockProcessor`] that processes the [`Block`] type.
+pub struct NomosBlockProcessor<ClPool, ClPoolAdapter, DaPool, DaPoolAdapter, RuntimeServiceId>
+where
+    ClPool: MemPool,
+    ClPool::Item: Clone + Eq,
+    ClPoolAdapter: MempoolNetworkAdapter<RuntimeServiceId>,
+    DaPool: MemPool,
+    DaPool::Item: Clone + Eq,
+    DaPoolAdapter: MempoolNetworkAdapter<RuntimeServiceId>,
+{
+    cl_mempool_relay: ClMempoolRelay<ClPool, ClPoolAdapter, RuntimeServiceId>,
+    da_mempool_relay: DaMempoolRelay<DaPool, DaPoolAdapter, DaPool::Key, RuntimeServiceId>,
+    sampling_relay: SamplingRelay<DaPool::Key>,
+    block_subscription_sender: broadcast::Sender<Block<ClPool::Item, DaPool::Item>>,
+}
+
+#[async_trait::async_trait]
+impl<ClPool, ClPoolAdapter, DaPool, DaPoolAdapter, RuntimeServiceId>
+    BlockProcessor<RuntimeServiceId>
+    for NomosBlockProcessor<ClPool, ClPoolAdapter, DaPool, DaPoolAdapter, RuntimeServiceId>
+where
+    ClPool: MemPool,
+    ClPool::Item: Transaction<Hash = ClPool::Key>
+        + Debug
+        + Clone
+        + Eq
+        + Serialize
+        + DeserializeOwned
+        + Send
+        + Sync
+        + 'static,
+    ClPool::Key: Send,
+    ClPoolAdapter: MempoolNetworkAdapter<RuntimeServiceId>,
+    DaPool: MemPool,
+    DaPool::Item: DispersedBlobInfo<BlobId = DaPool::Key>
+        + Debug
+        + Clone
+        + Eq
+        + Serialize
+        + DeserializeOwned
+        + Send
+        + Sync
+        + 'static,
+    DaPool::Key: Send + Ord + Debug,
+    DaPoolAdapter: MempoolNetworkAdapter<RuntimeServiceId>,
+{
+    type Block = Block<ClPool::Item, DaPool::Item>;
+
+    /// Try to add a [`Block`] to [`Cryptarchia`].
+    /// The updated [`Cryptarchia`] is returned if the block is valid and added
+    /// successfully. If not, [`Error`] is returned along with the unchanged
+    /// [`Cryptarchia`].
+    #[instrument(level = "debug", skip(self, cryptarchia, storage))]
+    async fn process_block<CryptarchiaState, Storage>(
+        &mut self,
+        cryptarchia: Cryptarchia<CryptarchiaState>,
+        block: Self::Block,
+        storage: &mut Storage,
+    ) -> Result<Cryptarchia<CryptarchiaState>, (Error, Cryptarchia<CryptarchiaState>)>
```

Comment:
`Cryptarchia` should be clonable, we don't need to return it in the event of an error